tuifimanager: init at 2.3.4

michaelBelsanti opened the pull request:
###### Description of changes
Packaged TUIFIManager and Unicurses
<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Commit:
tuifimanager: init at 2.3.4

Commit:
unicurses: init at 2.1.3

michaelBelsanti commented:
Removed a commit that wasn't meant to be included

michaelBelsanti commented on pkgs/development/python-modules/uni-curses/default.nix:
```diff
@@ -0,0 +1,31 @@
+  buildInputs = [ ncurses ];
+  propagatedBuildInputs = [ x256 ];
+
+  postPatch = ''
+    substituteInPlace './unicurses/__init__.py' \
+      --replace "find_library('ncursesw')" '"${ncurses}/lib/libncursesw.so.6"' \
+      --replace "find_library('panelw')" '"${ncurses}/lib/libpanelw.so.6"'
+  '';
```
Necessary because `ctypes.util.find_library` does not find the ncurses libraries. I couldn't find a cleaner fix, but if there is one, I'll modify it.

mattmelling replied:
> Necessary because ctypes.util.find_library does not find the ncurses libraries.

There seems to be a precedent for doing this [in other packages](https://github.com/NixOS/nixpkgs/blob/7deab23a756905fcbcc6fb13e7a27b0c0f3acdbf/pkgs/development/python-modules/pyzbar/default.nix#L26) so it should be fine, but please move this comment in to the code so it is easier to find.

michaelBelsanti commented on pkgs/applications/file-managers/tuifimanager/default.nix:
```diff
@@ -0,0 +1,29 @@
+  src = fetchFromGitHub {
+    repo = pname;
+    owner = "GiorgosXou";
+    rev = "1470b7b7d95fbf05548bbd60c8dacfc7fe4e9407";
```
Git repo doesn't have tags so I need to specify a commit. This commit refers to the commit "v.2.1.3"

SuperSandro2000 replied:
```suggestion
    # https://github.com/unicurses/unicurses/issues/29
    rev = "1470b7b7d95fbf05548bbd60c8dacfc7fe4e9407";
```

michaelBelsanti commented:
This will be a draft until #207996 is merged.

michaelBelsanti commented:
@ofborg eval

michaelBelsanti commented:
Result of `nixpkgs-review pr 208023` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>python310Packages.uni-curses</li>
    <li>python39Packages.uni-curses</li>
    <li>tuifimanager</li>
  </ul>
</details>

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/1655

mattmelling commented on pkgs/development/python-modules/uni-curses/default.nix:
```diff
@@ -0,0 +1,31 @@
+{ lib, buildPythonPackage, fetchPypi, ncurses, x256, setuptools }:
+
+buildPythonPackage rec {
+  pname = "unicurses";
```
pname is `unicurses` but attribute path is `uni-curses`, these should match

michaelBelsanti replied:
Whoops, corrected now.

mattmelling commented on pkgs/top-level/python-packages.nix:
```diff
@@ -11585,6 +11585,8 @@ self: super: with self; {
     unicorn-emu = pkgs.unicorn;
   };
 
+  uni-curses = callPackage ../development/python-modules/uni-curses { };
```
This does not match `pname` attribute in the derivation.

SuperSandro2000 replied:
```suggestion
  unicurses = callPackage ../development/python-modules/uni-curses { };
```

mattmelling commented on pkgs/applications/file-managers/tuifimanager/default.nix:
```diff
@@ -0,0 +1,29 @@
+  doCheck = false;
+
+  meta = with lib; {
+    description =
+      "A cross-platform terminal-based termux-oriented file manager (and component), meant to be used with a Uni-Curses project or as is. This project is mainly an attempt to get more attention to the Uni-Curses project";
```
```suggestion
    description = "A cross-platform terminal-based termux-oriented file manager";
    longDescription = ''
      A cross-platform terminal-based termux-oriented file manager (and component), 
      meant to be used with a Uni-Curses project or as is. This project is mainly an
      attempt to get more attention to the Uni-Curses project.
    '';
```

mattmelling commented on pkgs/applications/file-managers/tuifimanager/default.nix:
```diff
@@ -0,0 +1,29 @@
+{ lib, python3Packages, fetchFromGitHub }:
+
+python3Packages.buildPythonApplication rec {
+  pname = "tuifimanager";
+  version = "2.1.3";
+  format = "setuptools";
+
+  src = fetchFromGitHub {
+    repo = pname;
+    owner = "GiorgosXou";
+    rev = "1470b7b7d95fbf05548bbd60c8dacfc7fe4e9407";
+    sha256 = "sha256-nmfopFxtOE0zN0JOdESCU5ln4wShG6JzpjUXd6uvDgE=";
+  };
```
```suggestion
python3Packages.buildPythonApplication rec {
  pname = "tuifimanager";
  version = "2.1.9";
  format = "setuptools";

  src = fetchFromGitHub {
    repo = pname;
    owner = "GiorgosXou";
    rev = "v${version}";
    sha256 = "0i2yvvpmqpv7hwvsn83q16xdr7a0hv30w3jsvf2zxp2z2xr856s4";
  };
```

mattmelling replied:
Upstream has now tagged a 2.1.9 release.

SuperSandro2000 commented on pkgs/development/python-modules/uni-curses/default.nix:
```diff
@@ -0,0 +1,31 @@
+    sha256 = "sha256-Anz5YMZULdMhZ802T3tj9LOcBZZfUML1fPOJbC47nVA=";
+  };
+
+  buildInputs = [ ncurses ];
```
Does the program link against ncurses? If it is only used in postPatch then this line does nothing

michaelBelsanti replied:
You're right, it builds without it. Should this line be removed and ncurses only be referenced in the patch? Or should something be put in its place?

SuperSandro2000 replied:
The line can then be removed because it does nothing

SuperSandro2000 commented on pkgs/development/python-modules/uni-curses/default.nix:
```diff
@@ -0,0 +1,31 @@
+      --replace "find_library('ncursesw')" '"${ncurses}/lib/libncursesw.so.6"' \
+      --replace "find_library('panelw')" '"${ncurses}/lib/libpanelw.so.6"'
+  '';
+
```
Are there tests? If not add doCheck = false with a short comment.

michaelBelsanti replied:
I believe unicurses does have tests, but tuifi does not.

SuperSandro2000 commented on pkgs/applications/file-managers/tuifimanager/default.nix:
```diff
@@ -0,0 +1,33 @@
+    sha256 = "0i2yvvpmqpv7hwvsn83q16xdr7a0hv30w3jsvf2zxp2z2xr856s4";
+  };
+
+  propagatedBuildInputs = with python3Packages; [ uni-curses ];
```
```suggestion
  propagatedBuildInputs = with python3Packages; [ unicurses ];
```

michaelBelsanti commented:
I've updated both packages to the newest version. TUIFIManager added a new dependency, send2trash, which I can't get it to build with. It is already packages and I added it to propagatedBuildInputs, but it's failing to build with this error:
```
> ERROR: Could not find a version that satisfies the requirement Send2Trash==1.8.0 (from tuifimanager) (from versions: none)
> ERROR: No matching distribution found for Send2Trash==1.8.0
```

mattmelling commented:
The Send2Trash version in nixpkgs is `1.8.1b0`, whereas TUFIManager is looking for an exact match on version `1.8.0`. Take a look at the section "Using pythonRelaxDepsHook" in the [nixpkgs manual](https://nixos.org/manual/nixpkgs/stable/).

michaelBelsanti commented:
Oops, ignore that commit

michaelBelsanti commented:
> The Send2Trash version in nixpkgs is `1.8.1b0`, whereas TUFIManager is looking for an exact match on version `1.8.0`. Take a look at the section "Using pythonRelaxDepsHook" in the [nixpkgs manual](https://nixos.org/manual/nixpkgs/stable/).

Thanks, I gave that a look. I think I'm using it correctly, but I'm getting the same error. Do you see anything I'm missing?

mattmelling commented:
I got this to build by removing `pythonRelaxDepsHook` and patching `pyproject.toml` list this:

```
postPatch = ''
  substituteInPlace pyproject.toml \
    --replace "Send2Trash == 1.8.0" "Send2Trash >= 1.8.0"
'';
```

`tuifimanager` seems to specify the dependency in multiple places so maybe that is confusing `pythonRelaxDepsHook`.

michaelBelsanti commented:
Thank you! I hadn't had time to work on this more, but it seems you're right and the relaxDeps hook doesn't work with how the dependencies are specified in this project. Your snippet has it building and the binary works as expected.

Result of `nixpkgs-review pr 208023` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>python310Packages.unicurses</li>
    <li>python311Packages.unicurses</li>
    <li>tuifimanager</li>
  </ul>
</details>
